<a href="https://colab.research.google.com/github/MisterMirmix/LLM-Prompt-Engineering-RAG/blob/embeddings/LLM_Prompt_Engineering_RAG_%D0%9F%D1%80%D0%B0%D0%BA%D1%82%D0%B8%D0%BA%D0%B0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Тестовая работа с различными моделями и библиотеками

## sentence-transformers/all-MiniLM-L6-v2

In [2]:
from sentence_transformers import SentenceTransformer
sentences = ["This is an example sentence", "Each sentence is converted"]

model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')
embeddings = model.encode(sentences)
print(embeddings)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

[[ 6.76569194e-02  6.34959415e-02  4.87130806e-02  7.93049484e-02
   3.74481045e-02  2.65281182e-03  3.93749401e-02 -7.09844474e-03
   5.93614615e-02  3.15370038e-02  6.00980595e-02 -5.29052690e-02
   4.06068154e-02 -2.59308740e-02  2.98427939e-02  1.12690637e-03
   7.35148638e-02 -5.03818318e-02 -1.22386597e-01  2.37028114e-02
   2.97265425e-02  4.24768664e-02  2.56337449e-02  1.99515442e-03
  -5.69190457e-02 -2.71598250e-02 -3.29035148e-02  6.60249218e-02
   1.19007207e-01 -4.58791330e-02 -7.26214126e-02 -3.25840451e-02
   5.23412861e-02  4.50553298e-02  8.25300720e-03  3.67024392e-02
  -1.39415609e-02  6.53918684e-02 -2.64272150e-02  2.06383236e-04
  -1.36643508e-02 -3.62810828e-02 -1.95043813e-02 -2.89738141e-02
   3.94270159e-02 -8.84090737e-02  2.62426236e-03  1.36713963e-02
   4.83063012e-02 -3.11566219e-02 -1.17329165e-01 -5.11690676e-02
  -8.85288119e-02 -2.18963195e-02  1.42986281e-02  4.44167927e-02
  -1.34815322e-02  7.43392557e-02  2.66382676e-02 -1.98762883e-02
   1.79191

## вычисление косинусных и евклидовых расстояний

In [1]:
from sklearn.metrics.pairwise import cosine_distances, euclidean_distances

v1 = [[1, 2]]
v2 = [[5, 5]]

print("euclidean_distances", *euclidean_distances(v1, v2)[0])
print("cosine_distances", *cosine_distances(v1, v2)[0])

# euclidean_distances 5.0
# cosine_distances 0.05131670194948634

euclidean_distances 5.0
cosine_distances 0.05131670194948634


#Выполнение задач по практике

##3. Развертывание локальной модели эмбеддингов

**Задание:**
- Использовать модель `sentence-transformers/all-MiniLM-L6-v2` с помощью библиотеки Python sentence_transformers.
- Получить эмбеддинги для текста:
  - `"Привет, Мир!"`
  - `"Привет, ПММ!"`
- Рассчитать косинусное сходство между эмбеддингами.

**Ожидаемый результат:**
- Скрипт с результатами эмбеддингов и небольшое сравнение по косинусной мере.

In [19]:
from sentence_transformers import SentenceTransformer
model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')

# Получение эмбеддингов для текста "Привет, Мир!", "Привет, ПММ!"
sentences = ["Привет, Мир!", "Привет, ПММ!"]
embeddings = model.encode(sentences)

from sklearn.metrics.pairwise import cosine_distances
#Рассчет косинусного сходства между эмбеддингами.
cosine_similarity_matrix = 1. - cosine_distances(embeddings)

print(f"cosine_similarity_matrix:\n{cosine_similarity_matrix}")

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


cosine_similarity_matrix:
[[1.        0.8545017]
 [0.8545017 1.       ]]


In [20]:
print(f"embeddings:\n{embeddings}")

embeddings:
[[ 1.17196068e-02  4.93621156e-02  8.15440435e-03 -8.28170404e-03
  -8.07354003e-02  1.37737943e-02  1.29308298e-01  5.55282123e-02
   5.55311777e-02  4.64367354e-03  1.62926838e-02  2.82424167e-02
   2.40118541e-02  1.83201786e-02 -5.01855165e-02 -5.90683147e-02
  -2.36659329e-02  4.01300453e-02  4.84813452e-02  2.36179195e-02
  -3.40621062e-02 -8.29051584e-02  7.41608962e-02  2.18984205e-03
  -6.46943897e-02  6.99526519e-02 -1.58602726e-02  5.12738004e-02
   4.15975936e-02 -1.71495229e-02  2.88111549e-02  4.26895771e-04
   5.18569872e-02 -2.59287599e-02  3.97441462e-02  5.15948534e-02
  -4.00950871e-02 -7.37245232e-02  5.66229504e-03  5.80664948e-02
  -4.10499871e-02 -5.81693761e-02 -7.55470544e-02  9.30383727e-02
   4.15899828e-02  8.97833258e-02 -2.56140381e-02  3.57143693e-02
   4.66435682e-03  2.13995893e-02 -7.20800981e-02  5.20720892e-02
  -1.11638857e-02  2.49794894e-03  4.46381867e-02 -1.05812959e-01
   6.25862405e-02 -4.64097187e-02 -5.03809005e-03 -7.03720078e-0

Вывод:
*   Получены эмбеддинги для вышеуказанного текста
*   Была расчитана матрица сходства по косинусной мере, тексты "Привет, Мир!"
"Привет, ПММ!" являются схожими по косинусной мере, так как ее значение 0.85 приближенно к 1

